In [2]:
from sklearn.datasets import load_boston

boston = load_boston()
boston

{'data': array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
         4.9800e+00],
        [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
         9.1400e+00],
        [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
         4.0300e+00],
        ...,
        [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         5.6400e+00],
        [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
         6.4800e+00],
        [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         7.8800e+00]]),
 'target': array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
        18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
        15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
        13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
        21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
        35.4, 24.7, 3

In [3]:
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

# Regression dengan KNN

In [4]:
from sklearn import datasets
import pandas as pd

bstn_data = datasets.load_boston()
df_boston = pd.DataFrame(bstn_data.data,columns=bstn_data.feature_names)
df_boston['Target'] = pd.Series(bstn_data.target)
df_boston.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  Target  
0     15.3  396.90   4.98    24.0  
1     17.8  396.90   9.14    21.6  
2     17.8  392.83   4.03    34.7  
3     18.7  394.63   2.94    33.4  
4     18.7  396.90   5.33    36.2

In [5]:
import pandas as pd

rumah = {'age': [65.2, 78.9, 61.1, 45.8, 54.2], 
          'rad': [1, 2, 2, 3, 3],
          'harga': [24, 21.6, 34.7, 33.4, 36.2]}

rumah_df = pd.DataFrame(rumah)
rumah_df

age  rad  harga
0  65.2    1   24.0
1  78.9    2   21.6
2  61.1    2   34.7
3  45.8    3   33.4
4  54.2    3   36.2

- Dictionary menampung 3 keys yaitu : 'age', 'rad', 'harga'
- 'age' akan berasosiasi dengan nilai age rumah
- 'rad' akan berasosiasi dengan rad rumah
- 'harga' akan berasosiasi dengan sekumpulan nilai harga rumah
- rumah merupakan variabel yang menampung dictionary
- pd.DataFrame(rumah) untuk membentuk dictionary rumah menjadi pandas data frame
- objek dataframe ditampung dalam variabel rumah_df
- data age dan rad berperan sebagai features
- data harga berperan sebagai target

### Alasan memilih feature age dan rad

### Features & Target

In [6]:
import numpy as np

X_train = np.array(rumah_df[['age', 'rad']])
y_train = np.array(rumah_df['harga'])

print(f'X_train:\n{X_train}\n')
print(f'y_train: {y_train}')

X_train:
[[65.2  1. ]
 [78.9  2. ]
 [61.1  2. ]
 [45.8  3. ]
 [54.2  3. ]]

y_train: [24.  21.6 34.7 33.4 36.2]


- age dan rad di set kedalam np.array dan ditampung dalam variabel x_train sebagai features untuk training set
- harga di set ke dalam np.array dan ditampung dalam variabel y_train sebagai target untuk training set
- nilai features sudah termasuk format array dua dimensi dan tipe data numerik

### Training KNN Regression Model

In [7]:
from sklearn.neighbors import KNeighborsRegressor

K = 3
model = KNeighborsRegressor(n_neighbors=K)
model.fit(X_train, y_train)

KNeighborsRegressor(n_neighbors=3)

### Prediksi Harga Rumah

In [8]:
X_new = np.array([[61.1, 2.]])
X_new

array([[61.1,  2. ]])

- Prediksi harga rumah berdasarkan data age dan rad
- age 61.2 th
- rad 2
- [[61.1, 2.]] merupakan nilai feature yang akan di prediksi

In [9]:
y_pred = model.predict(X_new)
y_pred

array([31.63333333])

### Evaluasi KNN Regression Model

In [10]:
X_test = np.array([[62.2, 1], [76.9, 2], [59.1, 2], [41.8, 3], [50.2, 3]])
y_test = np.array([30.3, 25.7, 31.6, 33.7, 37.7])

print(f'X_test:\n{X_test}\n')
print(f'y_test: {y_test}')

X_test:
[[62.2  1. ]
 [76.9  2. ]
 [59.1  2. ]
 [41.8  3. ]
 [50.2  3. ]]

y_test: [30.3 25.7 31.6 33.7 37.7]


In [11]:
y_pred = model.predict(X_test)
y_pred

array([31.63333333, 26.76666667, 31.63333333, 34.76666667, 34.76666667])

- untuk rumah dengan age 62.2 dan rad 1 diprediksi memiliki harga 31.63333333, sedangkan data yang diharapkan adalah 30.3
- untuk rumah dengan age 76.9 dan rad 2 diprediksi memiliki harga 26.76666667, sedangkan data yang diharapkan adalah 25.7
- untuk rumah dengan age 59.1 dan rad 2 diprediksi memiliki harga 31.63333333, sedangkan data yang diharapkan adalah 31.6
- untuk rumah dengan age 41.8 dan rad 3 diprediksi memiliki harga 34.76666667, sedangkan data yang diharapkan adalah 33.7
- untuk rumah dengan age 50.2 dan rad 3 diprediksi memiliki harga 34.76666667, sedangkan data yang diharapkan adalah 37.7

#### Coefficient of Determination atau  𝑅2

In [12]:
from sklearn.metrics import r2_score

r_squared = r2_score(y_test, y_pred)

print(f'R-squared: {r_squared}')

R-squared: 0.8375399269906454


- Semakin nilai R-squared mendekati 1, semakin baik

#### Mean Absolute Error (MAE) atau Mean Absolute Deviation (MAD)

In [13]:
from sklearn.metrics import mean_absolute_error

MAE = mean_absolute_error(y_test, y_pred)

print(f'MAE: {MAE}')

MAE: 1.2866666666666682


- Semakin kecil nilai MAE akan mengindikasi kualitas modal yang lebih baik

#### Mean Squared Error (MSE) atau Mean Squared Deviation (MSD)

In [14]:
from sklearn.metrics import mean_squared_error

MSE = mean_squared_error(y_test, y_pred)

print(f'MSE: {MSE}')

MSE: 2.5317777777777835


- Semakin kecil nilai MSE akan mengindikasi kualitas modal yang lebih baik